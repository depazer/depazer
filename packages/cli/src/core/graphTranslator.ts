import type { ModuleGraph, ModuleObject } from '@/types/moduleGraph'

const graph: ModuleGraph = []

export function graphTranslator(
  moduleObject: ModuleObject,
  depth: number = 0,
  isDevDependency: boolean = true
): ModuleGraph {
  if (depth === 0) graph.length = 0

  /** @todo 标记开发依赖 */
  if (depth === 1) {
    isDevDependency = false
  }

  const name = moduleObject.name + '@' + moduleObject.version
  if (graph.find((item) => item.name === name) !== undefined) return graph

  graph.push({
    name,
    depth,
    isDevDependency,
    dependencies: moduleObject.dependencies.map((item) => item.name + '@' + item.version)
  })

  moduleObject.dependencies.forEach((item) => {
    graphTranslator(item, depth + 1, isDevDependency)
  })

  return graph
}
