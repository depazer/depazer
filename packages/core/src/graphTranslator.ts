import type { ModuleGraph, ModuleObject } from '@/types/moduleGraph'

const graph: ModuleGraph = []

export function graphTranslator(
  moduleObject: ModuleObject,
  depth: number = 0,
  isDevDependency: boolean = true
): ModuleGraph {
  if (depth === 0) {
    graph.length = 0
    isDevDependency = false
  }

  const name = generateModuleName(moduleObject)
  const isExist = graph.find((item) => item.name === name)

  if (isExist && (isExist.dependencies.length || !moduleObject.dependencies)) return graph

  if (isExist !== undefined) {
    isExist.dependencies = moduleObject.dependencies.map(generateModuleName)
    isExist.depth = Math.min(depth, isExist.depth)
  } else {
    graph.push({
      name,
      depth,
      isDevDependency,
      dependencies: moduleObject.dependencies.map(generateModuleName)
    })
  }

  if (depth === 0) {
    graph[0].dependencies.push(...moduleObject.devDependencies!.map(generateModuleName))
    graph[0].dependencies = [...new Set(graph[0].dependencies)]
  }

  moduleObject.dependencies.forEach((item) => graphTranslator(item, depth + 1, isDevDependency))
  moduleObject.devDependencies?.forEach((item) => graphTranslator(item, depth + 1))

  return graph
}

function generateModuleName(module: ModuleObject) {
  return module.name + '@' + module.version
}
