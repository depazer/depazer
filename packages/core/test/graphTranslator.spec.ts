import { graphTranslator } from '@/graphTranslator'
import { ModuleObject } from '@/index'
import { describe, test } from 'vitest'

describe('graphTranslator', () => {
  test('should translate a module object to a graph', ({ expect }) => {
    const moduleObject: ModuleObject = {
      name: 'a',
      version: '1.0.0',
      dependencies: [{ name: 'b', version: '1.0.0', dependencies: [] }],
      devDependencies: []
    }

    const graph = graphTranslator(moduleObject)
    expect(graph).toMatchSnapshot()
  })

  test('should translate a module object with devDependencies to a graph', ({ expect }) => {
    const moduleObject: ModuleObject = {
      name: 'a',
      version: '1.0.0',
      dependencies: [{ name: 'b', version: '1.0.0', dependencies: [] }],
      devDependencies: [{ name: 'c', version: '1.0.0', dependencies: [] }]
    }

    const graph = graphTranslator(moduleObject)

    expect(graph).toMatchSnapshot()
  })

  test('should translate a module object with loop / repetitive dependencies', ({ expect }) => {
    const moduleObject: ModuleObject = {
      name: 'a',
      version: '1.0.0',
      dependencies: [
        {
          name: 'c',
          version: '1.0.0',
          dependencies: [{ name: 'b', version: '1.0.0', dependencies: [] }]
        },
        {
          name: 'b',
          version: '1.0.0',
          dependencies: []
        }
      ],
      devDependencies: [
        {
          name: 'b',
          version: '1.0.0',
          dependencies: [{ name: 'a', version: '1.0.0', dependencies: [] }]
        }
      ]
    }

    const graph = graphTranslator(moduleObject)

    expect(graph).toMatchSnapshot()
  })
})
