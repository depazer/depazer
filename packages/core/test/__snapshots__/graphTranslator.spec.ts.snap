// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphTranslator > should translate a module object to a graph 1`] = `
[
  {
    "dependencies": [
      "b@1.0.0",
    ],
    "depth": 0,
    "isDevDependency": false,
    "name": "a@1.0.0",
  },
  {
    "dependencies": [],
    "depth": 1,
    "isDevDependency": false,
    "name": "b@1.0.0",
  },
]
`;

exports[`graphTranslator > should translate a module object with devDependencies to a graph 1`] = `
[
  {
    "dependencies": [
      "b@1.0.0",
      "c@1.0.0",
    ],
    "depth": 0,
    "isDevDependency": false,
    "name": "a@1.0.0",
  },
  {
    "dependencies": [],
    "depth": 1,
    "isDevDependency": false,
    "name": "b@1.0.0",
  },
  {
    "dependencies": [],
    "depth": 1,
    "isDevDependency": true,
    "name": "c@1.0.0",
  },
]
`;

exports[`graphTranslator > should translate a module object with loop / repetitive dependencies 1`] = `
[
  {
    "dependencies": [
      "c@1.0.0",
      "b@1.0.0",
    ],
    "depth": 0,
    "isDevDependency": false,
    "name": "a@1.0.0",
  },
  {
    "dependencies": [
      "b@1.0.0",
    ],
    "depth": 1,
    "isDevDependency": false,
    "name": "c@1.0.0",
  },
  {
    "dependencies": [
      "a@1.0.0",
    ],
    "depth": 1,
    "isDevDependency": false,
    "name": "b@1.0.0",
  },
]
`;
